{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import make_plots\n",
    "import data_prep_stack_barplots as prep\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from qiime2.plugins import feature_table\n",
    "from qiime2 import Artifact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def auto_qiime(directory,trimlength):\n",
    "    \"\"\"function to run auto qiime2 bash script, outputs data in appropriate form to work with for plotting\"\"\"\n",
    "    subprocess.run(['bash','-c','bash auto_qiime.sh '+directory+' '+trimlength])\n",
    "\n",
    "\"\"\"takes inputs of directory of .fastq files, trim length, and sampling depth for running the auto qiime script and creating quality plots\"\"\"\n",
    "directory=input('Directory of .fastq files:')\n",
    "trimlength=input('Sequencing trim length:')\n",
    "if trimlength.isdigit():\n",
    "    pass\n",
    "else:\n",
    "    raise TypeError('trim length input must be a positive integer')\n",
    "\n",
    "samp_depth = input(\"Desired sampling depth:\")\n",
    "if samp_depth.isdigit():\n",
    "    samp_depth = int(samp_depth)\n",
    "else:\n",
    "    raise TypeError('sampling depth must be a positive integer')\n",
    "\n",
    "\"\"\"runs auto_qiime function\"\"\"\n",
    "auto_qiime(directory,trimlength)\n",
    "\n",
    "\"\"\"read in newly created taxonomy data file to pandas\"\"\"\n",
    "taxonomy = pd.read_csv(\"data/taxonomy.tsv\", sep='\\t')\n",
    "taxonomy[['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']] = taxonomy['Taxon'].str.split(';', expand=True)\n",
    "taxonomy.set_index('Feature ID', inplace=True)\n",
    "\n",
    "\"\"\"reads in table.qza file from qiime2 into DataFrame\"\"\"\n",
    "unrarefied_table = Artifact.load('table.qza')\n",
    "rarefy_result = feature_table.methods.rarefy(table=unrarefied_table, sampling_depth=samp_depth)\n",
    "rarefied_table = rarefy_result.rarefied_table\n",
    "tabl = rarefied_table.view(pd.DataFrame)\n",
    "\n",
    "# add in any other data structures that need to be read in\n",
    "\n",
    "\n",
    "\"\"\"pre process data for relative abundance stacked bar plots\"\"\"\n",
    "kingdom_df, phylum_df, class_df, order_df, family_df, genus_df, species_df = prep.prepare_data_stacked_barplots(tabl, tax2)\n",
    "\n",
    "\"\"\"create plotly figures\"\"\"\n",
    "king_plot = make_plots.plotly_stacked_barplot(kingdom_df, 'Kingdom Relative Abundances')\n",
    "phy_plot = make_plots.plotly_stacked_barplot(phylum_df, 'Phylum Relative Abundances')\n",
    "class_plot = make_plots.plotly_stacked_barplot(class_df, 'Class Relative Abundances')\n",
    "ord_plot = make_plots.plotly_stacked_barplot(order_df, 'Order Relative Abundances')\n",
    "fam_plot = make_plots.plotly_stacked_barplot(family_df, 'Family Relative Abundances')\n",
    "gen_plot = make_plots.plotly_stacked_barplot(genus_df, 'Genus Relative Abundances')\n",
    "spec_plot = make_plots.plotly_stacked_barplot(species_df, 'Species Relative Abundances')\n",
    "\n",
    "qual_plot = make_plots.plot_qualities(directory, samp_depth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Taxon</th>\n",
       "      <th>Confidence</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4b5eeb300368260019c1fbc7a3c718fc</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>0.991607</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Bacteroidetes</td>\n",
       "      <td>c__Bacteroidia</td>\n",
       "      <td>o__Bacteroidales</td>\n",
       "      <td>f__Bacteroidaceae</td>\n",
       "      <td>g__Bacteroides</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fe30ff0f71a38a39cf1717ec2be3a2fc</th>\n",
       "      <td>k__Bacteria; p__Proteobacteria; c__Betaproteob...</td>\n",
       "      <td>0.982863</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Proteobacteria</td>\n",
       "      <td>c__Betaproteobacteria</td>\n",
       "      <td>o__Neisseriales</td>\n",
       "      <td>f__Neisseriaceae</td>\n",
       "      <td>g__Neisseria</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d29fe3c70564fc0f69f2c03e0d1e5561</th>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Firmicutes</td>\n",
       "      <td>c__Bacilli</td>\n",
       "      <td>o__Lactobacillales</td>\n",
       "      <td>f__Streptococcaceae</td>\n",
       "      <td>g__Streptococcus</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868528ca947bc57b69ffdf83e6b73bae</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>0.985898</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Bacteroidetes</td>\n",
       "      <td>c__Bacteroidia</td>\n",
       "      <td>o__Bacteroidales</td>\n",
       "      <td>f__Bacteroidaceae</td>\n",
       "      <td>g__Bacteroides</td>\n",
       "      <td>s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154709e160e8cada6bfb21115acc80f5</th>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>p__Bacteroidetes</td>\n",
       "      <td>c__Bacteroidia</td>\n",
       "      <td>o__Bacteroidales</td>\n",
       "      <td>f__Bacteroidaceae</td>\n",
       "      <td>g__Bacteroides</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                              Taxon  \\\n",
       "Feature ID                                                                            \n",
       "4b5eeb300368260019c1fbc7a3c718fc  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...   \n",
       "fe30ff0f71a38a39cf1717ec2be3a2fc  k__Bacteria; p__Proteobacteria; c__Betaproteob...   \n",
       "d29fe3c70564fc0f69f2c03e0d1e5561  k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...   \n",
       "868528ca947bc57b69ffdf83e6b73bae  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...   \n",
       "154709e160e8cada6bfb21115acc80f5  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...   \n",
       "\n",
       "                                  Confidence      kingdom              phylum  \\\n",
       "Feature ID                                                                      \n",
       "4b5eeb300368260019c1fbc7a3c718fc    0.991607  k__Bacteria    p__Bacteroidetes   \n",
       "fe30ff0f71a38a39cf1717ec2be3a2fc    0.982863  k__Bacteria   p__Proteobacteria   \n",
       "d29fe3c70564fc0f69f2c03e0d1e5561    1.000000  k__Bacteria       p__Firmicutes   \n",
       "868528ca947bc57b69ffdf83e6b73bae    0.985898  k__Bacteria    p__Bacteroidetes   \n",
       "154709e160e8cada6bfb21115acc80f5    1.000000  k__Bacteria    p__Bacteroidetes   \n",
       "\n",
       "                                                   class                order  \\\n",
       "Feature ID                                                                      \n",
       "4b5eeb300368260019c1fbc7a3c718fc          c__Bacteroidia     o__Bacteroidales   \n",
       "fe30ff0f71a38a39cf1717ec2be3a2fc   c__Betaproteobacteria      o__Neisseriales   \n",
       "d29fe3c70564fc0f69f2c03e0d1e5561              c__Bacilli   o__Lactobacillales   \n",
       "868528ca947bc57b69ffdf83e6b73bae          c__Bacteroidia     o__Bacteroidales   \n",
       "154709e160e8cada6bfb21115acc80f5          c__Bacteroidia     o__Bacteroidales   \n",
       "\n",
       "                                                family              genus  \\\n",
       "Feature ID                                                                  \n",
       "4b5eeb300368260019c1fbc7a3c718fc     f__Bacteroidaceae     g__Bacteroides   \n",
       "fe30ff0f71a38a39cf1717ec2be3a2fc      f__Neisseriaceae       g__Neisseria   \n",
       "d29fe3c70564fc0f69f2c03e0d1e5561   f__Streptococcaceae   g__Streptococcus   \n",
       "868528ca947bc57b69ffdf83e6b73bae     f__Bacteroidaceae     g__Bacteroides   \n",
       "154709e160e8cada6bfb21115acc80f5     f__Bacteroidaceae     g__Bacteroides   \n",
       "\n",
       "                                 species  \n",
       "Feature ID                                \n",
       "4b5eeb300368260019c1fbc7a3c718fc     s__  \n",
       "fe30ff0f71a38a39cf1717ec2be3a2fc    None  \n",
       "d29fe3c70564fc0f69f2c03e0d1e5561    None  \n",
       "868528ca947bc57b69ffdf83e6b73bae     s__  \n",
       "154709e160e8cada6bfb21115acc80f5    None  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax2 = pd.read_csv(\"872a0592-bb99-4296-941e-37dbbdff0d7a/data/taxonomy.tsv\", sep='\\t')\n",
    "tax2[['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']] = tax2['Taxon'].str.split(';', expand=True)\n",
    "tax2.set_index('Feature ID', inplace=True)\n",
    "tax2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
